### ISSUE TITLE
Objekte mit Formular wählen und filtern (QGIS 3.22/3.28)

### ISSUE THREAD

**Issue by ackmar** [opengisch/QGIS-Issues-SH#112]
Das Filtern zwischen zwei Datümer (Typ QDate) ergibt falsches Ergebnis.
("Ausstellungsdatum">='01.01.2022' AND "Ausstellungsdatum"<='31.12.2022')

Wird wahrscheinlich nicht als Datum erkannt und funktioniert leider nur per Korrektur mit folgendem Ausdruck.
("Ausstellungsdatum">=to_date('2022-01-01') AND "Ausstellungsdatum"<=to_date('2022-12-31'))

![image](https://user-images.githubusercontent.com/34340243/207785889-050b766d-ca52-4964-8bab-c56eb297cb89.png)

**Comment by signedav** [comment]
Ja, ich denke das ist so. Ich glaube zwar, dass dies gehen würde:
("Ausstellungsdatum">='2022-01-01' AND "Ausstellungsdatum"<='2022-12-31')

Dennoch ist das to:date stabiler. Dort kannst du auch das Format angeben so: 

("Ausstellungsdatum">=to_date('01.01.2022','dd.MM.yyyy') AND "Ausstellungsdatum"<=to_date('31.12.2022','dd.MM.yyyy')

**Comment by ackmar** [comment]
Leider filtert diese Formel keine Daten und gibt wieder den gesamten Inhalt aus :
("Ausstellungsdatum">='2022-01-01' AND "Ausstellungsdatum"<='2022-06-30')

![image](https://user-images.githubusercontent.com/34340243/208031618-81dc6fd3-b088-46da-a298-772127265334.png)

Kann man dies nicht als Bug deklarieren? Der User erwartet natürlich aus dem oben definierten Filter eine korrekte Auswertung und nicht erst bei der nachträglichen Modifizierung in der unteren Ausdruckszeile.

**Comment by signedav** [comment]
Du übergibst halt einen String. Ich dachte das "amerikanische" Format wird irgendwie erkannt aber offenbar nicht. Ein Bug ist es meiner Meinung nach nicht. Aber natürlich schon etwas umständlich so.

**Comment by signedav** [comment]
Du benutzt die Filterfunktion im GUI und die generiert eine "falsche" Expression.

_Sorry, jetzt begreif ichs. Habs am Handy nicht gesehen (Nachteil, wenn man beim Frühstück am Handy Issue trackt. Response time top - Qualität flop. Wenigstens gratis ;-))_

Ja, das scheint mir schon ein Bug zu sein.

**Comment by ackmar** [comment]
_Ich hätte es auch klarer formulieren können. Das tut Eurer super Qualität nichts ab._

Könntest Du bitte an Romedi eine Aufwandschätzung für das Bug Fixing mitteilen. Je nach Grösse des Aufwandes würden wir den Bug gerne beheben lassen.

Vorab besten Dank.

**Comment by signedav** [comment]
Also das hier würde ja gehen `"Ausstellungsdatum">='2021-09-19' AND "Ausstellungsdatum"<='2022-12-19'`.

Das heisst irgendwo hier https://github.com/qgis/QGIS/blob/master/src/gui/editorwidgets/qgsdatetimesearchwidgetwrapper.cpp#L86-L97 müssten wir das `DateFormat` mitgeben. Etwa so:
```
return fieldName + "='" + v.toString(Qt::ISODate) + '\'';
```

Vielleicht ist es das schon. Allerdings wird hier von QVariant konvertiert. Ist das immer `QDateTime`? :thinking: Ich mach mal eine Schätzung.

**Comment by signedav** [comment]
Btw. hier der Issue auf QGIS https://github.com/qgis/QGIS/issues/43177

**Comment by ackmar** [comment]
Gegenüber dem QDate wird es mit QDateTime etwas komplexer. Die Zeit oder und Zeitzone ist im nachstehenden Filter nicht unterstützt.

Bsp. ("test">='2022-12-19 05:00:00' AND "test"<='2022-12-21 20:00:00')

**Comment by signedav** [comment]
Kannst du das testen: ("test">='2022-12-19T05:00:00' AND "test"<='2022-12-21T20:00:00')

**Comment by ackmar** [comment]
So würde es klappen.

**Comment by signedav** [comment]
Hab mir das gerade codemässig angeschaut und etwas bemerkt, dass mir vorhin noch nicht aufgefallen ist. Der Wert wird immer in das Feldformat konvertiert. Was hast du dort eingetragen?

![image](https://user-images.githubusercontent.com/28384354/222170150-cc16a647-06c9-4b36-a8a6-94f34938e2af.png)

Sofern das Feld ein `QDate` ist, sollte dort eigentlich immer `yyyy-MM-dd` drin stehen. Kannst du das mal prüfen @ackmar ?

**Comment by signedav** [comment]
Zu meiner Info:

Das Format wird bereits ermittelt. Und zwar hier https://github.com/qgis/QGIS/blob/master/src/gui/editorwidgets/qgsdatetimesearchwidgetwrapper.cpp#L49-L59 Anhand des genannten Settings. Ich werd mal noch in die Implementierung dieses Settings schauen. Eventuell ist es eher für den Fall, dass das Feld kein spezifischer Datumstyp hat...

**Comment by signedav** [comment]
Dieses Setting ist aber wahrlich verwirrend. Ein PR sollte diese Verwirrung lösen und Fehlkonfigurationen vermeiden.

https://github.com/qgis/QGIS/pull/52063

**Comment by ackmar** [comment]
Jetzt bin ich total verwirrt. In meinem Beispiel sieht das Setting folgendermassen aus:

![image](https://user-images.githubusercontent.com/34340243/222341552-610734dd-160a-41af-bdf0-99dc7a933cb0.png)

![image](https://user-images.githubusercontent.com/34340243/222341721-535461c5-f4a4-4430-bca1-1ec13beaec5f.png)

Das Feldformat und die Customanzeige sind hier identisch und wurden aus einem SQL mit folgender Umformung definiert:
to_char(e.erstellt_am :: DATE, 'dd.MM.yyyy') as "Erstellungsdatum"

Als User würde ich in diesem Beispiel mindestens das Format erwarten, in dem es ursprünglich definiert ist, aber hier muss trotzdem das amerikanische Format für den Filter angewendet werden.

Wird die angebliche Lösung für 3.30 auch auf 3.28 backported?

**Comment by signedav** [comment]
Ich denke nicht dass es gebackportet wird, da es nur hilft Fehlkonfigurationen zu vermeiden. 

Aber dein Fall hier scheint noch etwas anderes zu benötigen. Ich schau mal weiter und melde mich.

**Comment by signedav** [comment]
Habn noch was gefunden

Habe ich ein `date` column in der SQL Tabelle, lädt mir QGIS 3.28 das schweizer Format in die Settings (anhand meiner Systemsprache):
![image](https://user-images.githubusercontent.com/28384354/222385006-0ede93d2-4120-454f-9268-70858bc20ae6.png)

Dies ist gefixt worden (vor zwei Wochen) für 3.30 https://github.com/qgis/QGIS/pull/51471

Hier wird das technische Datumsformat beibehalten und die Anzeige gemäss Systemsprache angepasst:
![image](https://user-images.githubusercontent.com/28384354/222385463-1736eb89-8fc0-4871-a4c0-b856abbd63bf.png)

Ich denke, das würde auch dein Problem lösen. Dennoch schau ich noch dein SQL Case an...

**Comment by signedav** [comment]
Dein SQL Case kann ich hingegen nicht reproduzieren:
![image](https://user-images.githubusercontent.com/28384354/222388878-5eca4c8c-a95d-45c6-9bdc-7dc10e6984dc.png)

Ich denke aber dass wir folgendermassen weiterfahren können:

1. Konfiguriere dein Field Format zu "yyyy-MM-dd" die Anzeige kannst du auf dem schweizer Format lassen. Filter sollten nun funktionieren.
2. Zukünftig (ab 3.30) wird nun auch das korrekte Field Format beim erstellen des Layers genommen.

Falls euch ein Backport wichtig ist, könnte ich das mal anfragen. Allerdings würde ich gerne darauf verzichten, da der Bug nicht zu grossen Problemen führt. Ausserdem wäre die Config deines aktuellen Projektes ja falsch gesetzt. Die müsste man auch mit dem Fix noch umkonfigurieren...

**Comment by ackmar** [comment]
@signedav 
Es funktioniert wie Du im Abschnitt 1 beschrieben hast. Wir verzichten auf die Anfrage mit dem Backport und schauen dann mit der künftigen LTR-Version im nächsten Jahr. Die Sache muss daher nicht mehr weiterverfolgt werden.

Herzlichen Dank für deine Abklärungen und deinen Service.

**Comment by ZsanettMed** [comment]
kein ZOHO ticket, warten auf die Version 3.30
